---
title: "Class 14"
author: "Yaniv Iny (PID: A18090586)"
format: pdf
toc: true
---

## Background

The data for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

## Data Import
```{r}
counts <- read.csv("GSE37704_featurecounts.csv", row.names =1 )

coldata <-read.csv("GSE37704_metadata.csv")
```



## Inspect and tidy data

Does the `counts` columns math with the `colData`rows?

```{r}
head(counts)
```

```{r}
head(coldata)
```

```{r}
coldata$id
```
```{r}
colnames(counts)
```
The fix here would be to remove the first "length" column.

```{r}
countdata <- counts[,-1]
head(countdata)
```

Check for matching countdata and coldata

```{r}
colnames(countdata) == coldata$id
```

> Q1. How many genes total 

```{r}
nrow(countdata)
```


> Q2. Filter to remove zero count genes (rows where there are zero counts in all comuns).

```{r}
to.keep.inds <- rowSums(countdata) > 0
```

```{r}
 new.counts <-countdata[to.keep.inds,]
```

```{r}
nrow(new.counts)
```

## Setup for DESeq 

```{r}
#/ message: false 
library(DESeq2)
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = new.counts,
                             colData = coldata,
                             design = ~condition)
```

## Run DESeq
```{r}
dds <- DESeq(dds)
res <- results(dds)
```
```{r}
head(res)
```

## Volcano plot of results 

```{r}
library(ggplot2)
```
```{r}
ggplot(res)+
  aes(log2FoldChange, -log(padj))+
  geom_point()+
  geom_vline(xintercept = c(-2,2), col="red") +
  geom_hline(yintercept = 2, col = "red")
```

##Gene annotation
Add Gene SYMBOL and ENTREZID
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=rownames(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL")
                 

res$entrez = mapIds(org.Hs.eg.db,
                    keys=rownames(res),
                    keytype="ENSEMBL",
                    column="ENTREZID")
                  


```

##Pathway Analysis

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
library(pathview)

```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)

```

```{r}
head(keggres$less)

```
Cell cycle figure

```{r}
pathview(foldchanges, pathway.id = "hsa04110")
```
Insert this figure in my report

![](hsa04110.pathview.png)


```{r}
pathview(foldchanges, pathway.id = "hsa03030")
```

![](hsa03030.pathview.png)

```{r}
pathview(foldchanges, pathway.id = "hsa03440")

```
![](hsa03440.pathview.png)


## Gene ontology
```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

```{r}
head(gobpres$less)
```

